### PR TITLE
add quirk system & full support for the G305

### PR DESCRIPTION
**Pull request by FFY00**

### CODE REVIEW

**Comment by bentiss** [comment]
Don't you want to say *add*, not *ass* :)

**Comment by whot** [comment]
EDONKEY :)

File: src/hidpp20.c [hunk 1]
```diff
@@ -2620,6 +2620,16 @@ hidpp20_onboard_profiles_initialize(struct hidpp20_device *device,
 						  HIDPP20_USER_PROFILES_G402,
 						  profiles->sector_size,
 						  data);
+
+	if (device->quirk != HIDPP20_QUIRK_G305)
```
Comment (bentiss) [inline src/hidpp20.c]: Don't you want to check on `rc` first?

Comment (FFY00) [reply src/hidpp20.c]: Yeah :man_facepalming:

File: src/hidpp20.h [hunk 2]
```diff
@@ -74,6 +77,12 @@ hidpp20_device_new(const struct hidpp_device *base, unsigned int idx);
 void
 hidpp20_device_destroy(struct hidpp20_device *device);
 
+/* -------------------------------------------------------------------------- */
+/* quirks                                                                     */
+/* -------------------------------------------------------------------------- */
+
+#define HIDPP20_QUIRK_NONE				0
+
```
Comment (whot) [inline src/hidpp20.h]: better to use enums here, gives us the illusion of type safety

File: src/libratbag-data.c [hunk 3]
```diff
@@ -166,6 +166,11 @@ init_data_hidpp20(struct ratbag *ratbag,
 
 	str = g_key_file_get_string(keyfile, group, "Quirk", NULL);
 	data->hidpp20.quirk = HIDPP20_QUIRK_NONE;
+	if (str)
+	{
+		if (streq(str, "G305"))
+			data->hidpp20.quirk = HIDPP20_QUIRK_G305;
+	}
```
[the remark this answers is not in the document]
Comment (FFY00) [reply src/libratbag-data.c]: The idea was to add other quirks bellow, so I would prefer to keep it this way.

**Comment by FFY00** [comment]
@whot

File: src/libratbag-data.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (whot) [reply src/libratbag-data.c]: That's fine when your patchset includes the next quirks but AIUI we don't have any other quirks on the immediate horizon, right? So let's change this to the best code right now and alter it when or if it is required in the future.

And besides, I'd still need you to put the `{` on the previous line anyway ;)

[the remark this answers is not in the document]
Comment (whot) [reply src/libratbag-data.c]: oh, that is already done in the last version, so let's get this merged then, to tired today to wait for this.

**Comment by whot** [comment]
See #786 please, I had to revert this one.

**Comment by FFY00** [comment]
Oh, this did not rebase properly.

File: src/libratbag-data.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (FFY00) [reply src/libratbag-data.c]: We need a quirk for the G602 so yeah, there is another quirk in the horizon.

**Pull request merged**